### PR TITLE
cluster: add focus weights (in favor of #26)

### PR DESCRIPTION
**Pull request by fneum**
This PR offers a different implementation for `focus_weights` which allows to set a proportion of the total number of clusters for particular countries. The rest (countries not listed) is distributed according to mean load. Defaults to no focus weights.

In the ``config.yaml`` add for instance:

```yaml
focus_weights:
    DE: 0.9
    FR: 0.1
```

which results in
focus_weights

The sum of focus weights must be less than 1.

### CODE REVIEW

File: scripts/cluster_network.py [hunk 1]
```diff
@@ -160,6 +162,22 @@ def distribute_clusters(n, n_clusters, solver_name=None):
...
         "Number of clusters must be {} <= n_clusters <= {} for this selection of countries.".format(len(N), N.sum())
 
+    if focus_weights is not None:
+
+        total_focus = sum(list(focus_weights.values()))
+
+        assert total_focus <= 1.0, "The sum of focus weights must be less than 1."
+        
+        for country, weight in focus_weights.items():
+            L[country] = weight
```
Comment (FabianHofmann) [inline scripts/cluster_network.py]:
```suggestion
            L[country] = weight / len(L[country])
```
Like this you don't have to worry about how many subnetworks lie in the country and L always sums up to zero. So the assertion check below is not necessary

File: scripts/cluster_network.py [hunk 2]
```diff
@@ -160,6 +162,22 @@ def distribute_clusters(n, n_clusters, solver_name=None):
...
+
+        remainder = [c not in focus_weights.keys() for c in L.index.get_level_values('country')]
+        L[remainder] = L.loc[remainder].pipe(normed) * (1 - total_focus)
+
+        logger.warning('Using custom focus weights for determining number of clusters.')
+
+    assert L.sum() == 1.0, "Country weights L must sum up to 1.0 when distributing clusters."
```
Comment (FabianHofmann) [inline scripts/cluster_network.py]:
```suggestion
```

File: scripts/cluster_network.py [hunk 3]
```diff
@@ -160,6 +162,22 @@ def distribute_clusters(n, n_clusters, solver_name=None):
     assert n_clusters >= len(N) and n_clusters <= N.sum(), \
         "Number of clusters must be {} <= n_clusters <= {} for this selection of countries.".format(len(N), N.sum())
 
+    if focus_weights is not None:
+
+        total_focus = sum(list(focus_weights.values()))
+
+        assert total_focus <= 1.0, "The sum of focus weights must be less than 1."
```
Comment (FabianHofmann) [inline scripts/cluster_network.py]:
```suggestion
        assert total_focus <= 1.0, "The sum of focus weights must be less or equal 1."
```

File: scripts/cluster_network.py (hunk 1 above)
Comment (fneum) [reply scripts/cluster_network.py]: No, the idea is for the focus weight is to specify the share of clusters to be allocated to that country (e.g. DE: 0.5 should result in 50% of the nodes allocated to Germany and the remainder is distributed proportional to mean load for all the other countries which are not listed). Your suggestion would distort this.

**Comment by fneum** [comment]
Thanks for reviewing :+1:

File: scripts/cluster_network.py (hunk 1 above)
Comment (FabianHofmann) [reply scripts/cluster_network.py]: Okay, quick question: 
What if Denmark has 2 synchronous zones and we want to give Denmark 40% of all nodes. Then `L["DK"]` would include two entries. When we give them both a value of 0.4, the total of Denmark would have a weight of 80%. By dividing by `len(L["DK"])`, we guarantee that the resulting total weight of denmark is 40%. 

Am I missing a point here?

Comment (fneum) [reply scripts/cluster_network.py]: Ah, I was confused because of the assertion check which should still be done, but you are right.

Comment (fneum) [reply scripts/cluster_network.py]: changed in https://github.com/PyPSA/pypsa-eur/commit/7df2b76f9daeb8d2775b833edf486c770e46c682

**Pull request merged**